ggu/main.c

```diff
@@ -16,11 +16,337 @@
+	entry.u.ggu.params.renewal_step_cycle =
+		rte_be_to_cpu_64(policy->params.renewal_step_cycle);
+
+	send_gk_cmd(block_id, &entry, gk_conf);
+}
+
+static void
+process_single_packet(struct rte_mbuf *pkt, struct ggu_config *ggu_conf)
+{
+	uint8_t j;
+	uint8_t *data;
+	uint8_t *policy_ptr;
+	uint8_t num_policies;
+	uint16_t *proto_ptr;
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr *ip4hdr;
+	struct ipv6_hdr *ip6hdr;
+	struct udp_hdr *udphdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+		ip4hdr = rte_pktmbuf_mtod_offset(pkt, 
+			struct ipv4_hdr *, sizeof(struct ether_hdr));
+		if (ip4hdr->next_proto_id != IPPROTO_UDP) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received non-UDP packets, ntuple filter bug!\n");
+			goto free_packet;
+		}
+
+		udphdr = (struct udp_hdr *)&ip4hdr[1];
+		break;
+
+	case ETHER_TYPE_IPv6:
+		ip6hdr = rte_pktmbuf_mtod_offset(pkt, 
+			struct ipv6_hdr *, sizeof(struct ether_hdr));
+		if (ip6hdr->proto != IPPROTO_UDP) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received non-UDP packets, ntuple filter bug!\n");
+			goto free_packet;
+		}
+
+		udphdr = (struct udp_hdr *)&ip6hdr[1];
+		break;
+
+	default:
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: unknown network layer protocol %hu\n",
+			ether_type);
+		goto free_packet;
+		break;
+	}
+
+	if (udphdr->src_port != rte_cpu_to_be_16(ggu_conf->ggu_src_port) ||
+			udphdr->dst_port !=
+			rte_cpu_to_be_16(ggu_conf->ggu_dst_port)) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: unknown udp src port %hu, dst port %hu, ntuple filter bug!\n",
+			rte_be_to_cpu_16(udphdr->src_port),
+			rte_be_to_cpu_16(udphdr->dst_port));
+		goto free_packet;
+	}
+
+	data = (uint8_t *)&udphdr[1];
+	num_policies = data[0];
+	data = data + 1;
+	proto_ptr = (uint16_t *)data;
+	policy_ptr = (uint8_t *)((uint16_t *)data + num_policies);
+
+	/* Loop over each policy decision on the packet. */
+	for (j = 0; j < num_policies; j++) {
+		int ret;
+		struct ggu_policy policy;
+
+		ret = policy_parse_next(proto_ptr, policy_ptr, &policy);
+		if (ret < 0)
+			goto free_packet;
+
+		proto_ptr = proto_ptr + 1;
+		if (policy.flow.proto == ETHER_TYPE_IPv4)
+			policy_ptr = policy_ptr + sizeof(policy.flow.f.v4) +
+				sizeof(policy.params);
+		else
+			policy_ptr = policy_ptr + sizeof(policy.flow.f.v6) +
+				sizeof(policy.params);
+
+		process_single_policy(&policy, ggu_conf);
+	}
+
+free_packet:
+	rte_pktmbuf_free(pkt);
+}
+
+static int
+ggu_proc(void *arg)
+{
+	int ret;
+	uint32_t lcore = rte_lcore_id();
+	struct ggu_config *ggu_conf = (struct ggu_config *)arg;
+	uint8_t port_in = ggu_conf->net->back.id;
+	uint16_t rx_queue = ggu_conf->rx_queue_front;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"ggu: the GK-GT unit is running at lcore = %u\n", lcore);
+
+	/* Wait for network devices to start. */
+	while (ggu_conf->net->configuring)
+		;
+
+	while (ggu_conf->net->front.rss_configured == false)
```

Comment:
The reason is that RSS should be configured after the NICs start, otherwise, the RSS configuration will be reset once the NICs start.